### ISSUE TITLE
Binary Output Should Be Names as such

### ISSUE THREAD

**Issue by jcrubino** [rusthon/Rusthon#25]
All binaries should be named as such.  
   e.g. main.go from the example.md is a little confusing for text editors

**Comment by hartsantler** [comment]
main.go is a bug, it should be main-go.  and others should be main-c++ and main-rust.  perhaps if the result of the markdown is just a single exe, then it can simply be called `main`.

In the future we need to have a way for the user to name their exe's from the markdown.

**Comment by jcrubino** [comment]
will fix

**Comment by kud1ing** [comment]
Currently names like `rusthon-*-build` are hard-coded. How about using the markdown file name?

**Comment by hartsantler** [comment]
kud1ing is right, It makes sense to use the markdown file name.

**Comment by Shoozza** [comment]
This would only work if you have only one md file, or one script file.
What if there are move files?

**Comment by kud1ing** [comment]
What is our multi-file story anyway? If we generated multilpe C*+ files, is the C++ runtime generated inside every cpp file or only once externally and then included?

**Comment by hartsantler** [comment]
What do you guys think the most common use case is going to be?

My guess is that most of the time, the user just needs a single exe that is built and linked by Rusthon, so all the c++ code can simply be dumped into one big file.  

On the other hand, the more advanced user may not want Rusthon to build and link the final exe, they have some other build system like cmake, and they just want to use Rusthon to generate the project with multiple files and proper headers.  At some point we are going to need to define a special markup for the fenced code blocks for this, maybe like: `@filename:foobar` which would then generate foobar.cpp and foobar.hpp

**Comment by kud1ing** [comment]
My instinct is that the simple case should stay simple and the advanced stuff should be made possible.
Ideally the two use-cases should not interfer with each other.

If Rusthon had a Java backend, it would be very interesting for people doing iOS/Android development. But that would need some more deep thinking/refactoring. Anyway i enjoy having this idea in the back of my mind.

**Comment by Shoozza** [comment]
A few ideas how this could be handled:

1.  Optionally use the first comment in the code block `// filename: main.cpp` to determine the file name for a block
1.  Same as 1 but add special markup ``@filename:main.cpp`` in front of a code block instead
1.  Put all blocks in their corresponding rusthon-language-build.ext file (one file per language)
1.   Same as 3 but per md file basis: test.md -> test.md.cpp, test2.md -> test2.md.cpp

For non md files the filenames could be used (they shouldn't collide with the md file names due to the .md extension in the generated file name for md files. See point 4.)

**Comment by hartsantler** [comment]
@kud1ing there is a plan for a Java frontend, so imagine you import something like libGDX into rusthon, and then recompile it as C++, and run it on both iOS and Android.  Then we can skip doing a Java backend.  Although a Java backend is probably still nice to have, and doing it might not be all that difficult from this point.  However, at the moment, and probably for the next three months, my focus is entirely on the c++ backend.

@Shoozza everything your saying looks right on.  I guess the other thing we need to think about is what is the syntax for linking external libs that were maybe compiled before hand by the user, `libsomething.a`.  Do we allow the user to trigger external builds from the markdown, do we allow custom gcc and linker flags?  It might be nice to allow automatic building of external libs, because that can also call some wrapper generator, and rusthon could also parse the headers to gain knowledge about the classes and functions in there using something like https://pypi.python.org/pypi/CppHeaderParser/

**Comment by hartsantler** [comment]
this bug is fixed in the Pythia fork of Rusthon.
https://github.com/secureosv/pythia
